Add ADR 0010 for Release Artifacts

lranjbar opened the pull request:
Signed-off-by: Lisa Rashidi-Ranjbar <lisa.a.rashidi-ranjbar@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->

- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [X] Documentation content changes
- [ ] Other... Please describe:

## What has been updated?

Adding ADR to define release artifacts for EdgeX Foundry.

lranjbar commented:
> @lranjbar - I think this is good list of what is release artifact. Do we need to suggest which of these are tightly coupled to a freeze and release. For example, I think the compose files and docs are independent. But we didn't like it when the tests changed independent of the code. Can we add a section (or otherwise annotate) what things are frozen and released together and what is not.

This is a good addition... I went ahead and added a note of which of these are tied to a code release. For the tests I left this as undecided because it does seem tied to the release of code to me in the same way the docker compose files are.

> To that end, we don't speak to SDKs and app/device services. Would tests around app/device services be tied to those independent releases?

I'm not sure if we have thought about this... and it is a good point to bring up. From my perspective is that it seems that most of our tests are against the services in edgex-go. I think it makes sense to have targeted release tests for device services if we have them as well. We will want to figure out how to specify the tests against different repositories at a point in time. We could add more tags that are descriptive... or something else I'm not sure.

anonymouse64 commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,84 @@
+
+*Tied to Code Release?* Yes
+
+Snaps are released for every named release of EdgeX Foundry. During development the community stages snaps to the `latest/edge` channel of the [Snapcraft Store](https://snapcraft.io). At the time of code freeze we will promote the snaps from `latest/edge` to the `latest/beta` channels as beta release. At the time of release we will promote the snaps from `latest/beta` to `latest/stable`. In addit ...
```
this could use a comment that there will be/should be additional testing done of the snap in latest/beta before it goes to latest/stable and <release-track>/stable

anonymouse64 commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,84 @@
+
+*Tied to Code Release?* Yes
+
+Snaps are released for every named release of EdgeX Foundry. During development the community stages snaps to the `latest/edge` channel of the [Snapcraft Store](https://snapcraft.io). At the time of code freeze we will promote the snaps from `latest/edge` to the `latest/beta` channels as beta release. At the time of release we will promote the snaps from `latest/beta` to `latest/stable`. In addit ...
```
also what was the decision re: versioning and tagging the various snap releases so that the version field in the snap matches the release tag? I.e. so we don't end up with `version: v1.1.1-dev.2` in the snap on latest/stable ? 

I remember suggesting that the release job download the built snap artifacts on the latest/beta channel, unpack and change the version number, then re-upload and release to the latest/stable channel.

lranjbar replied:
During the dot release we had more issues with releasing the dev versions not only in snaps but in the docker images as well. We found that the version api endpoint for the services was returning the dev versions in both. To fix this in the dot release I did a manual rebuild of everything. 

For Hanoi I'm going to make an update to the automation ADR as well. Essentially at code freeze we will need to rebuild the release candidates with the proper `version (ie: v1.1.x)`. For snaps this means at beginning of code freeze we will push that candidate to the `beta/candidate` channels for integration testing. At the end of the code freeze we should be able to promote that candidate to the `stable` channel. 

For docker images it's similar because we are doing an image promotion from `docker.staging` to `docker.release+Dockerhub` at release time. So during code freeze we will have to build those images and push them to staging to get ready for the release.

This means that during code freeze we will have to rebuild the artifacts each time we fix found issues too.

anonymouse64 replied:
sounds good to me, thanks

MightyNerdEric commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,84 @@
+
+*Tied to Code Release?* Yes
+
+Docker images are released for every named release of EdgeX Foundry. During development the community releases images to the `docker.staging` repository in [Nexus](http://nexus3.edgexfoundry.org). At the time of release we promote the last tested image from `docker.staging` to `docker.release`. In addition to that we will publish the docker image on [DockerHub](http://hub.docker.com).
```
```suggestion
Docker images are released for every named release of EdgeX Foundry. During development the community releases images to the `docker.staging` repository in [Nexus](http://nexus3.edgexfoundry.org). At the time of release we promote the last tested image from `docker.staging` to `docker.release`. In addition to that we will publish the docker image on [DockerHub](https://hub.docker.com/orgs/edgexfoundry).
```

lranjbar commented:
Update 7/2/2020:

I've added in a section that lists out our internal build dependencies. I've also implemented the changes requested.

Cloud I'd like to mark the testing frameworks as linked to the code release. In my opinion it is similar to the docker compose files. After the release we will release a tag for the testing frameworks. Any feedback you have would be great. @cloudxxx8

cloudxxx8 commented:
Yes, @lranjbar , I agree with that.  From the release point of view, they are just like docker compose files.  Thanks for letting me know.

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+## Context
+
+During the Geneva release of EdgeX Foundry the DevOps WG transformed the CI/CD process with new Jenkins pipeline functionality. After this new functionality was added we also started adding release automation. This new automation is outlined in ADR 0007 Release Automation. However, in ADR 0007 only three release artifact types are outlined. This document is meant to be a living document to try to ...
```
**nit**: add link to ADR 0007

tonyespy replied:
**suggest**: This document is meant to be a living document which outlines all currently supported artifacts associated with an EdgeX release, and should be updated if/when this list changes.

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+*Tied to Code Release?* Yes
+
+Docker images are released for every named release of EdgeX Foundry. During development the community releases images to the `docker.staging` repository in [Nexus](http://nexus3.edgexfoundry.org). At the time of release we promote the last tested image from `docker.staging` to `docker.release`. In addition to that we will publish the docker image on [DockerHub](https://hub.docker.com/orgs/edgexfo ...
```
One thing that's confusing in this ADR is that there actually are multiple project releases from EdgeX beyond the bi-annual "named" releases. Examples include:

 * SDKs
 * device services
 * application services

Some of these also result in release artifacts (i.e. app & device services) and some don't (SDKs).

lranjbar replied:
The goal is not to call out which repositories release each artifact type because that is called out in ADR 0007. Instead is just to define the artifact types and the general processes. Device and application services follow this same pattern for docker images the only difference is the schedule.

tonyespy replied:
Sure, but this section says that "Docker images are released for every named release of EdgeX Foundry."

My point is that they're also released for application and device service releases which are not "named".

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+##### docker.snapshots
+Retention Policy: 90 days since last download
+
+Contains: Docker images that are not expected to be released. This will contain EdgeX that images are built with each merge and should be available to test things against a specific commit. In addition to this we also build images to optimize the builds in the CI infrastructure.
```
**suggest**: These images are built for each project git merge (edgex-go, app services, device services) and are made available to developers can test images that correlate to individual commits. These images are also used to optimize builds in the CI infrastructure.

ernestojeda replied:
@tonyespy The images for edgex-go, app-services, device-services go into **staging** on each git merge to master. So we would probably move this below.

mkbhanda replied:
s/can test/so that they can test

For for my education, @lranjbar how do these images help optimise builds in CI infra?

lranjbar replied:
At the time this was written images were built with a git commit are currently pushed to both docker.staging and docker.snapshots. I will remove those sentences because it only seems to contain the ci-build-images now.

The images built from the repository [edgexfoundry/ci-build-images](https://github.com/edgexfoundry/ci-build-images)are stored in docker.snapshots in the Nexus. These images are made to optimize the builds.

@tonyespy @ernestojeda @mkbhanda

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+Docker Tags Used: Git commit, Version-Git commit, Version (ci-build-images only), Latest (ci-build-images-only)
+
+##### docker.staging
+Retention Policy: 180 days since last download
```
Any reason we're keeping these for longer than the snapshots? 6 months seems like a long time to be keeping release candidates around...

ernestojeda replied:
The LF uses staging as the holding repo before docker images are pushed to release or DockerHub

mkbhanda replied:
100 days is more like 30 months .. but from last download -- is that a user download of the image, this re-setting some freshness flag? Would this open up an attack vector ... periodically try to download an image to eventually run out of disk space?

lranjbar replied:
This is the repository that we would use to store potential release candidates really the time is just to take of the case of images that get less action. IE: Somebody updates the code a month into the release cycle and it could be deleted before release 4-5 months later.

The Linux Foundation hasn't really brought up the problem of disk space with us.. but in general the images in Nexus don't get much action towards them except for the images used by the testing frameworks.

tonyespy replied:
I was just asking the question because 6 months seemed like a long time. I'm fine with this as is then...

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+Docker Tags Used: Version (ie: v1.x), Release Branch (master, fuji, etc), Latest
+
+##### docker.release
+Retention Policy: No automatic removal. Requires TSC approval to remove images from this repository.
```
So are barcelona images still available? Should we be cleaning these up?

ernestojeda replied:
TSC can vote to remove any old images. For example you can look at all the versions tagged for `docker-core-config-seed-go` in the release repo here: https://nexus3.edgexfoundry.org/#browse/browse:docker.release:v2%2Fdocker-core-config-seed-go%2Ftags

mkbhanda replied:
these are good to keep around, useful for historical purposes including testing for regressions.

lranjbar replied:
When we wrote the Nexus policy we thought it should be up to the TSC to decide whether or not to clean up old releases.

jpwhitemn replied:
I have the issue of Docker (and Snap) image retention and purging as an Architects' meeting subject.  I think this ADR can proceed while we address this issue.

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+*Tied to Code Release?* Yes
+
+Docker compose files are released alongside the docker images for every release of EdgeX Foundry. During development the community maintains a compose file named `nightly-build`. This compose file is meant to be a beta version of EdgeX. At the time of release the community makes a named release compose file. (ie: `geneva`)
```
I'm not sure what you mean by "this compose file is meant to be a beta version of EdgeX"? It's not really meant as a beta (which is concept that some projects/products to release an early preview of a release in order to solicit customer/user feedback**), it's meant to be the compose file used during the entire development cycle.

****** we should discuss at some point whether we should be producing actual release candidates and/or Beta releases

tonyespy replied:
Also, do we have rules on when & how released compose files can be modified?

lranjbar replied:
We've only ever updated released compose files tied to a release or a dot release. Otherwise it doesn't seem to get much action and there isn't any rules around it today.

I realize beta is a loaded term, but the crux of it is that those compose files are used by the testing frameworks to run tests every night. Because of that it means it is tracking the code that is in master branch of the services. Which makes it  more of a "beta"/"experimental"/"dev" release of EdgeX. I could swap it out for experimental is that is more clear.

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+*Tied to Code Release?* Yes
+
+Docker compose files are released alongside the docker images for every release of EdgeX Foundry. During development the community maintains a compose file named `nightly-build`. This compose file is meant to be a beta version of EdgeX. At the time of release the community makes a named release compose file. (ie: `geneva`)
```
Also right now, there isn't a single compose file per release, there are multiple compose files (although Lenny's working to fix that).

lranjbar replied:
I updated the verbiage on this one to be more accurate.

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+### Github Page: EdgeX Docs
+
+*Tied to Code Release?* No
```
"Tied to Code Release" is a bit ambiguous, maybe should define it at the top of the ADR? The docs page actually does have a version drop-down, so while it's not tied to the release of source code, it is loosely tied to a specific release.

lranjbar replied:
I added this from the feedback from Jim... It just means to highlight the things that need to be released when we decide to do a code release. Not sure a better way to word it honestly.

jpwhitemn replied:
How about tied to an EdgeX major or minor release?

tonyespy replied:
I think what it means is whether or not the artifact is tied to an actual source code release, and in the case of edgex-docs, it's not. When a release happens, we create a new set of docs, but the docs don't write themselves when the code is released, a bunch of us instead update the docs asynchronously.

That said, it still might be a good idea to define this at the top of the document.

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+*Tied to Code Release?* No
+
+EdgeX Foundry releases a set of documentation for our project at http://docs.edgexfoundry.org. This page is a github page that is managed by the `edgex-docs` Github repository. As a community we make our best effort to keep these docs up to date. On this page we are also versioning the docs with the semantic versions of the named releases. As a community we try to version our documentation site s ...
```
**suggest**: These pages are generated from markdown hosted in the [edgex-docs](https://github.com/edgexfoundry/edgex-docs) repository.

**Note** - it'd be nice to add links in this doc to referenced repositories or docs.

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+*Tied to Code Release?* Yes, for the final semantic version
+
+Github tags are used to track the releases of EdgeX Foundry. During development the tags are incremented automatically using a development suffix (ie: `v1.1.1-dev.1` -> `v1.1.1-dev.2`). At the time of release we release a tag with the final semantic version (ie: `v1.1.1`).
```
**suggest**: are incremented automatically **for each commit**

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+*Tied to Code Release?* Yes
+
+Snaps are released for every named release of EdgeX Foundry. During development the community stages snaps to the `latest/edge` channel of the [Snapcraft Store](https://snapcraft.io/search?q=edgexfoundry). At the time of code freeze we will promote the snaps from `latest/edge` to the `latest/beta` amd `latest/candidate` channels as beta release. This beta release is to trigger validation of the r ...
```
Again, see previous comment about app and device services.

lranjbar replied:
So the goal is to try to keep the process the same or similar as much as possible. So that the only difference for those device services and application services is the timing of things. Even though certain things either get released at different times device services usually get a named version at some point. Application services have maintained the same schedule so far.. for the most part.

tonyespy replied:
Again, we try to release updated device and app services in sync with a named release, however that doesn't mean we released "named versions" of them (nor application services). This was done on purpose. The rule is that that the major release number of an application or device service must match the corresponding major version of a core EdgeX release to be compatible. So in theory, we should be able to run any 1.x version of a device service against a 1.x version of EdgeX. 

I agree the goal is to keep the process the same as much as possible too, but as worded it implies that we only ever release snaps when a named release happens and that's not 100% true. Same as with docker images above.

tonyespy replied:
Maybe just add a note that affect in both sections?

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+*Tied to Code Release?* Yes
+
+Snaps are released for every named release of EdgeX Foundry. During development the community stages snaps to the `latest/edge` channel of the [Snapcraft Store](https://snapcraft.io/search?q=edgexfoundry). At the time of code freeze we will promote the snaps from `latest/edge` to the `latest/beta` amd `latest/candidate` channels as beta release. This beta release is to trigger validation of the r ...
```
"During development the community stages snaps"

**suggest**: "During development the CI pipelines automatically publish daily builds of the snaps to..."

tonyespy replied:
**suggest**:

At the time of code freeze we will promote the snaps from `latest/edge` to the `latest/beta` **amd (sic)** `latest/candidate` channels as a release candidate, this triggers an additional set of automatic validation tests of the snap. In addition we also create a named release track for the release (ie: `geneva/stable`). At the time of release we will promote the snaps from `latest/beta` to the `latest/stable` and `<release>/stable` channels.

lranjbar replied:
The daily snap thing is actually only an edgex-go thing. The CI pipelines for the device services snaps push to the edge channel on every merge to master so it lands much faster for those repositories.

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+*Tied to Code Release?* No
+
+In addition to our documentation site EdgeX foundry also releases our API specifications on Swaggerhub.
```
link?

lranjbar replied:
Funny enough.. I actually don't know the link to our SwaggerHub. Is it listed anywhere?

@cloudxxx8 @bill-mahoney

jpwhitemn replied:
https://app.swaggerhub.com/search?owner=EdgeXFoundry1

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+*Tied to Code Release?* Yes
+
+The EdgeX Foundry community has a set of tests we maintain to do regression testing during development this framework is tracking the `master` branch of the components of EdgeX. At the time of release we will update the testing frameworks to point at the released Github tags and add a version tag to the testing frameworks themselves. This creates a snapshot of testing framework at the time of rel ...
```
suggest: ...testing during development**.**

Also I'm not sure what "This framework is tracking the `master` branch of the components of EdgeX" means? I thought that our blackbox tests run against docker images? When we run blackbox tests, we don't rebuild EdgeX and create docker images dynamically.

lranjbar replied:
On every merge to master there are docker images created with the docker tag, `master`. The testing frameworks are using these images to test so that they are in sync with the changes in the `master` branch.

tonyespy replied:
Thanks for the explanation. Maybe that should get folded into your description here, as it wasn't clear to me?

tonyespy commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+![Known EdgeX Foundry Build Dependencies](0010/known-build-dependencies.png)
+
+ - Edgex-go Snap: Has dependencies on app-service-configurable, device-virtual-go and support-rulesengine because they are included in the snap.
```
You can remove **support-rulesengine** from the snap.

lranjbar replied:
This was based off the Geneva release... did we do that already?

tonyespy replied:
It's been as part of Hanoi, so merged into master already...

lenny-intel commented on docs_src/design/adr/devops/0010-Release-Artifacts.md:
```diff
@@ -0,0 +1,95 @@
+
+![Known EdgeX Foundry Build Dependencies](0010/known-build-dependencies.png)
+
+ - Edgex-go Snap: Has dependencies on app-service-configurable, device-virtual-go and support-rulesengine because they are included in the snap.
```
Do we what to include all the go-mod-<name> dependencies?

lranjbar replied:
Go module dependencies are included at build time. This is to call out the fact that you have to build these in a specific order or you have to rebuild them all several times... that would not be the case with the go modules.

lranjbar commented:
Everyone, thanks for all the feedback! I've made updates to a lot of things from the suggestions. :)
